### PR TITLE
`Programming exercises`: Synchronize result display in exercise view

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Currently the result history and the result string for the newest result used different stylings for the icons and text color.

### Description

This PR unifies these two views and refactors the used methods to reduce code duplication.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise

1. Log in to Artemis
2. Navigate to the exercise and participate as a student
3. Check that in different configurations (befor/after/without due date, different assessment types, ...) and verify that the results are styled the same.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| result.utils.ts | 72% | 87% |

### Screenshots

Before (Red Cross vs. question mark)
Bildschirmfoto 2022-10-24 um 16 28 08
After (Both question marks)
Bildschirmfoto 2022-10-24 um 16 19 17

### CODE REVIEW

**Review by ge65cer — CHANGES_REQUESTED** [review]
Missing icon in the timeline for manual assessment.
Screenshot 2022-10-26 at 09 27 08

https://artemis-test1.artemis.in.tum.de/courses/62/exercises/1106 (Test User 1)

File: src/test/javascript/spec/component/overview/result-history/result-history.component.spec.ts [hunk 1]
```diff
@@ -96,20 +96,4 @@ describe('ResultHistoryComponent', () => {
         expect(component.showPreviousDivider).toBeTrue();
         expect(component.movedLastRatedResult).toBeTrue();
     });
-
-    it('should return the right values for result score', () => {
```
Comment (ge65cer) [inline src/test/javascript/spec/component/overview/result-history/result-history.component.spec.ts:100]: Unused imports.

File: src/main/java/de/tum/in/www1/artemis/domain/Feedback.java [hunk 2]
```diff
@@ -346,6 +346,24 @@ public boolean isStaticCodeAnalysisFeedback() {
         return this.text != null && this.text.startsWith(STATIC_CODE_ANALYSIS_FEEDBACK_IDENTIFIER) && this.type == FeedbackType.AUTOMATIC;
     }
 
+    /**
+     * Checks whether the feedback was created by a submission policy
+     * @return true if it is static code analysis feedback else false
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/domain/Feedback.java:351]: Does the retur text make sense here? Seems to be from another method

File: src/main/java/de/tum/in/www1/artemis/domain/Feedback.java [hunk 3]
```diff
@@ -346,6 +346,24 @@ public boolean isStaticCodeAnalysisFeedback() {
...
+     */
+    @JsonIgnore
+    public boolean isSubmissionPolicyFeedback() {
+        return this.text != null && this.text.startsWith(SUBMISSION_POLICY_FEEDBACK_IDENTIFIER) && this.type == FeedbackType.AUTOMATIC;
+    }
+
+    /**
+     * Checks whether the feedback was created by an automatic test
+     * @return true if it is static code analysis feedback else false
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/domain/Feedback.java:360]: Does the retur text make sense here? Seems to be from another method

File: src/main/webapp/app/exercises/shared/result/result.component.ts [hunk 4]
```diff
@@ -246,82 +196,10 @@ export class ResultComponent implements OnInit, OnChanges {
...
-            } else if (initializedResultWithScore(this.result) && (!assessmentDueDate || assessmentDueDate.isBefore(dayjs()))) {
-                // Submission is not in due time of exercise, has a result with score and there is no assessmentDueDate for the exercise or it lies in the past.
-                // TODO handle external submissions with new status "External"
-                return ResultTemplateStatus.LATE;
-            } else {
-                // Submission is not in due time of exercise and there is actually no feedback for the submission or the feedback should not be displayed yet.
-                return ResultTemplateStatus.LATE_NO_FEEDBACK;
-            }
+        if (this.showIcon) {
+            console.log(this.result);
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/exercises/shared/result/result.component.ts:201]: Console.log left in?

File: src/test/javascript/spec/component/utils/result.utils.spec.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+import { Result } from 'app/entities/result.model';
+
+describe('ResultUtils', () => {
+    it('should filter out all non unreferenced feedbacks', () => {
+        const feedbacks = [{ reference: 'foo' }, { reference: 'foo', type: FeedbackType.MANUAL_UNREFERENCED }, { type: FeedbackType.MANUAL_UNREFERENCED }, {}];
+        const unreferencedFeedbacks = getUnreferencedFeedback(feedbacks);
+        expect(unreferencedFeedbacks).toEqual([{ type: FeedbackType.MANUAL_UNREFERENCED }]);
+    });
+
+    it.each([
```
Comment (stefanwaldhauser) [inline src/test/javascript/spec/component/utils/result.utils.spec.ts:18]: <3 I like the it.each. Need to use it more also

File: src/test/javascript/spec/component/utils/result.utils.spec.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+import { FeedbackType, STATIC_CODE_ANALYSIS_FEEDBACK_IDENTIFIER } from 'app/entities/feedback.model';
+import { SubmissionExerciseType } from 'app/entities/submission.model';
+import { AssessmentType } from 'app/entities/assessment-type.model';
+import { ParticipationType } from 'app/entities/participation/participation.model';
+import { MIN_SCORE_GREEN, MIN_SCORE_ORANGE } from 'app/app.constants';
+import { faCheckCircle, faQuestionCircle, faTimesCircle } from '@fortawesome/free-regular-svg-icons';
+import { ExerciseType } from 'app/entities/exercise.model';
+import { Result } from 'app/entities/result.model';
+
+describe('ResultUtils', () => {
```
Comment (stefanwaldhauser) [inline src/test/javascript/spec/component/utils/result.utils.spec.ts:11]: Great Test!

**Review by stefanwaldhauser — DISMISSED** [review]
Works as expected except the small issues from above and some small things.

**Review by stefanwaldhauser — APPROVED** [review]
Approved since feedback was adressed! Gj :)

**Review by stefanwaldhauser — DISMISSED** [review]
Reapprove! Changes from last review are good! Gj

**Review by manuelmanso — CHANGES_REQUESTED** [review]
Code changes look good, however I found three issues when testing.

1. For text exercises, the submission view doesn't show the graded result even when there is one
image

2. For modelling exercises, the submission view has an issue displaying the date

image

3. Before starting a practice repository for a programming exercise you didn't participate in, the result shows "You have missed the deadline". After you start the practice repository, it shows "No graded result" as shown below
image

**Review by manuelmanso — APPROVED** [review]
Tested in ts1, issues have been fixed!

**Review by tobias-lippert — APPROVED** [review]
Code LGTM

**Review by simon-hng — DISMISSED** [review]
Code looks good

**Review by stefanwaldhauser — APPROVED** [review]
Reapprove!

**Review by tobias-lippert — APPROVED** [review]
Tested again on ts1. Everything worked as expected. Code looks still good to me.

**Pull request merged**